### PR TITLE
Pass Assets Directory in CI

### PR DESCRIPTION
**Pull request by scbedd**
The reason that .NET is seeing such inconsistent failures with

> The line cannot be parsed

Is due to two factors 

- The .NET tests run the collected _test libraries_ in parallel, even if within each they are running in serial.
- The test-proxy doesn't maintain a file lock on its breadcrumb file during an operation

Due to these reasons, we are having multiple processes writing to the `.breadcrumb` file at the root of the folder containing all our assets slices. When this occurs, sometimes there is a chance to end in a malformed file.

```txt
sdk/storage/Azure.Storage.Blobs/assets.json;mZpWxY1sKn;net/storage/Azure.Storage.Blobs_b46f524eaf
nt_a1e3a847ee
```

^ An example of what a file looked like at the end of a build.

This PR addresses the issue by ensuring that each test library invocation maintains its own folder of assets slices. This change only applies to CI, so local users will remain unaffected.

The intent will be to **remove** this code when we integrate the proxy that supports file locking on this breadcrumb file.

### CODE REVIEW

**Comment by scbedd** [comment]
This should be all the tests relying on `core`, which should include storage, but I've manually queued a [net - storage - ci](https://dev.azure.com/azure-sdk/public/_build/results?buildId=2951549&view=results) build.

File: sdk/core/Azure.Core.TestFramework/src/TestProxy.cs [hunk 1]
```diff
@@ -81,6 +81,11 @@ private TestProxy(string proxyPath, bool debugMode = false)
                 }
             };
 
+            if (!string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("TF_BUILD")))
```
Comment (scbedd) [inline sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:84]:
```suggestion
            bool.TryParse(Environment.GetEnvironmentVariable("TF_BUILD"), out bool inCI);

            if (inCI)
```

File: sdk/core/Azure.Core.TestFramework/src/TestProxy.cs [hunk 2]
```diff
@@ -81,6 +81,13 @@ private TestProxy(string proxyPath, bool debugMode = false)
                 }
             };
 
+            bool.TryParse(Environment.GetEnvironmentVariable("TF_BUILD"), out bool inCI);
```
Comment (scbedd) [inline sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:84]: image

Comment (jsquire) [reply sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:84]: Most of our build-related scripts use `$ci` as a flag.   Maybe we should consider naming this `CI_BUILD` for consistency?

Comment (christothes) [reply sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:84]: This env variable is defined by the CI host environment, so I don't think we have a choice on naming.

File: sdk/core/Azure.Core.TestFramework/src/TestProxy.cs [hunk 3]
```diff
@@ -81,6 +81,13 @@ private TestProxy(string proxyPath, bool debugMode = false)
...
+
+            if (inCI)
+            {
+                testProxyProcessInfo.EnvironmentVariables["PROXY_ASSETS_FOLDER"] = Path.Combine(System.IO.Path.GetTempPath(), Guid.NewGuid().ToString());
```
Comment (christothes) [inline sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:88]: Just curious - does this solve the problem by ensuring each proxy instance is working from a unique file path?  Where else is this environment variable referenced?

Comment (scbedd) [reply sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:88]: This is exactly correct. The issue is that all the proxy processes are accessing the same `.breadcrumb` file and we're seeing all the weird issues in .NET runs due to that. I didn't originally intend the proxy to be used that way, but I can't argue that we can support it!

Comment (scbedd) [reply sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:88]: > Where else is this environment variable referenced?

This is a [proxy specific variable.](https://github.com/Azure/azure-sdk-tools/blob/e02eeee926c971471a434062ca85a78c855f5eed/tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitAssetsConfiguration.cs#L95) Your framework doesn't need to care about it at all.

**Comment by JoshLove-msft** [comment]
If we think the more robust fix is in the proxy by locking the files (which I suppose would prevent this issue from occurring in other languages), then I'm not sure it is worth having an interim fix now that we know what the root cause is. Or is there a concern that it will take significantly longer to get the proxy fix in?

**Comment by scbedd** [comment]
> If we think the more robust fix is in the proxy by locking the files (which I suppose would prevent this issue from occurring in other languages), then I'm not sure it is worth having an interim fix now that we know what the root cause is. Or is there a concern that it will take significantly longer to get the proxy fix in?

Definitely worth the interim fix. Just save your devs the frustration IMO.

File: sdk/core/Azure.Core.TestFramework/src/TestProxy.cs (hunk 2 above)
Comment (JoshLove-msft) [inline sdk/core/Azure.Core.TestFramework/src/TestProxy.cs:84]: The changes from this PR can be reverted when https://github.com/Azure/azure-sdk-tools/issues/6562 is fixed and deployed.

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**